mesa/agent.py

```diff
@@ -57,3 +63,274 @@ def advance(self) -> None:
+        interactions with the model's environment and other agents.The implementation uses a WeakKeyDictionary to store agents,
+        which means that agents not referenced elsewhere in the program may be automatically removed from the AgentSet.
+    """
+
+    def __init__(self, agents: Iterable[Agent], model: Model):
+        """
+        Initializes the AgentSet with a collection of agents and a reference to the model.
+
+        Args:
+            agents (Iterable[Agent]): An iterable of Agent objects to be included in the set.
+            model (Model): The ABM model instance to which this AgentSet belongs.
+        """
+        self.model = model
+
+        if not self.model.agentset_experimental_warning_given:
+            self.model.agentset_experimental_warning_given = True
+            warnings.warn(
+                "The AgentSet is experimental. It may be changed or removed in any and all future releases, including patch releases.\n"
+                "We would love to hear what you think about this new feature. If you have any thoughts, share them with us here: https://github.com/projectmesa/mesa/discussions/1919",
+                FutureWarning,
+                stacklevel=2,
+            )
+
+        self._agents = weakref.WeakKeyDictionary()
+        for agent in agents:
+            self._agents[agent] = None
+
+    def __len__(self) -> int:
+        """Return the number of agents in the AgentSet."""
+        return len(self._agents)
+
+    def __iter__(self) -> Iterator[Agent]:
+        """Provide an iterator over the agents in the AgentSet."""
+        return self._agents.keys()
+
+    def __contains__(self, agent: Agent) -> bool:
+        """Check if an agent is in the AgentSet. Can be used like `agent in agentset`."""
+        return agent in self._agents
+
+    def select(
+        self,
+        filter_func: Callable[[Agent], bool] | None = None,
+        n: int = 0,
+        inplace: bool = False,
+        agent_type: type[Agent] | None = None,
+    ) -> AgentSet:
+        """
+        Select a subset of agents from the AgentSet based on a filter function and/or quantity limit.
+
+        Args:
+            filter_func (Callable[[Agent], bool], optional): A function that takes an Agent and returns True if the
+                agent should be included in the result. Defaults to None, meaning no filtering is applied.
+            n (int, optional): The number of agents to select. If 0, all matching agents are selected. Defaults to 0.
+            inplace (bool, optional): If True, modifies the current AgentSet; otherwise, returns a new AgentSet. Defaults to False.
+            agent_type (type[Agent], optional): The class type of the agents to select. Defaults to None, meaning no type filtering is applied.
+
+        Returns:
+            AgentSet: A new AgentSet containing the selected agents, unless inplace is True, in which case the current AgentSet is updated.
+        """
+        agents = list(self._agents.keys())
```

Comment:
Maybe I am overly cautious with respect to performance implications, but this function creates at worst 4 lists of agents, when we actually don't need any list, but just an iterable to pass to Agentset. I think we could use a generator to only iterate once and only up to n agents

```python
    def agent_generator():
        count = 0
        for agent in self:
            if filter_func and not filter_func(agent):
                continue
            if agent_type and not isinstance(agent, agent_type):
                continue
            yield agent
            count += 1
            if n is not None and count >= n:
                break

    agents = agent_generator()
```

The code might be a bit more difficult to understand, but I also think it might be worth it.

Reply:
`if n is not None and count >= n:` should be  `if n and count >= n`

Otherwise, I agree with the suggestions. 

@EwoutH  and me both independently added support for selecting on AgentType. The current implementation adds it as an additional argument, while my implementation replaced filter_func with key, where key was either a callable or agent type. The reason for my suggested implementation was to mimic the sort method signature and partially avoid multiple criteria for selection. I am fine either way.

Reply:
Did your implementation allow both an AgentType and a callable? Or would the preferred usage be calling `.select()` twice?

Reply:
It was either or, and deliberately so, because of the point made by @Corvince (which is now solved).

In my view, if you have multiple criteria on which to select, it is cleaner to implement them all together into a single selection function. This, by the way would avoid also multiple calls to select, which really is not desirable from a performance point.

Reply:
Great, we align in that regard. Will mark this as resolved.